app/models/compute_resources/foreman/model/openstack.rb

```diff
@@ -92,12 +94,37 @@ def boot_from_volume(args = {})
       } ]
     end
 
+    def possible_scheduler_hints
+      SEARCHABLE_ACTIONS.collect{|x| x.to_s.camelize }
+    end
+
+    def get_server_groups(policy)
+      server_groups = client.server_groups.select{ |sg| sg.policies.include?(policy) }
+      errors.add(:scheduler_hint_value, _("No matching servergroups found")) if server_groups.empty?
+      server_groups
+    end
+
+    def format_scheduler_hint_filter(args = {})
+      raise "Hint data is missing" if args[:scheduler_hint_data].nil?
+      name = args.delete(:scheduler_hint_filter).underscore.to_sym
+      data = args.delete(:scheduler_hint_data)
+      filter = {}
+      case name
+        when :server_group_anti_affinity, :server_group_affinity
+          filter[:group] = data[:scheduler_hint_value]
+        when :raw
+          filter = JSON.parse(data[:scheduler_hint_value])
+      end
+      args[:os_scheduler_hints] = filter
+    end
+
     def create_vm(args = {})
       boot_from_volume(args) if Foreman::Cast.to_bool(args[:boot_from_volume])
       network = args.delete(:network)
       # fix internal network format for fog.
       args[:nics].delete_if(&:blank?)
       args[:nics].map! {|nic| { 'net_id' => nic } }
+      format_scheduler_hint_filter(args) if args[:scheduler_hint_filter] != '' && !args[:scheduler_hint_filter].nil?
```

Comment:
`if args[:scheduler_hint_filter] != '' && !args[:scheduler_hint_filter].nil?` is equivalent to `if args[:scheduler_hint_filter].present?` which is a bit more readable and shorter